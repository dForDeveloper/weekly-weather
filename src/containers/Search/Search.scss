.Search {
  margin: 32px;
  height: 32px;
  text-align: center;
}

.Search--input {
  padding: 8px 0 0;
  font-size: 24px;
  outline: none;
  text-align: center;
  width: 320px;
  border: none;
  border-bottom: 2px solid #dfe1e4;

  &::placeholder {
    font-size: 20px;
    color: #dfe1e4;
  }

  &:focus {
    border-bottom: 2px solid $blue;
  }
}