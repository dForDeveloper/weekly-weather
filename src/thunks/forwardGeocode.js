import { fetchData } from '../utils/api';
import { setCoordinates, setCity, setError } from '../actions';

export const forwardGeocode = (query) => {
  return async (dispatch) => {
    query = query.replace(/\s/g, '+');
    query = query.replace(/,/g, '%2C');
    try {
      const url = `http://localhost:3001/api/v1/forwardgeocode/${query}`;
      const geocodeData = await fetchData(url);
      const { lat: latitude, lng: longitude } = geocodeData[0].geometry;
      const { city, state_code } = geocodeData[0].components;
      dispatch(setCity(`${city}, ${state_code}`));
      dispatch(setCoordinates({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}