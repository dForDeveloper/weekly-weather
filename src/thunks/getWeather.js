import { fetchData } from '../utils/api';
import { cleanData } from '../utils/dataCleaner';
import { setWeather, setError, toggleLoading } from '../actions';

export const getWeather = ({ latitude: lat, longitude: lon }) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const url = `http://localhost:3001/api/v1/weather/${lat}/${lon}`;
      const weather = await fetchData(url);
      dispatch(setWeather(cleanData(weather)));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message))      
    }
    dispatch(toggleLoading(false));
  }
}