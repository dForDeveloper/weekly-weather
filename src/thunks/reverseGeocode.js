import { fetchData } from '../utils/api';
import { setCity, setError } from '../actions';

export const reverseGeocode = ({ latitude: lat, longitude: lon }) => {
  return async (dispatch) => {
    try {
      const url = `http://weekly-weather.herokuapp.com/api/v1/reversegeocode/${lat}/${lon}`;
      const geocodeData = await fetchData(url);
      const { city, state_code } = geocodeData[0].components;
      dispatch(setCity(`${city}, ${state_code}`));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}