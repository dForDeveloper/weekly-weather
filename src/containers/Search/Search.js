import React, { Component } from 'react';
import { connect } from 'react-redux';
import { forwardGeocode } from '../../thunks/forwardGeocode';
import PropTypes from 'prop-types';

export class Search extends Component {
  constructor() {
    super();
    this.state = {
      query: ''
    }
  }

  handleChange = (event) => {
    this.setState({ query: event.target.value });
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    let { query } = this.state;
    this.props.forwardGeocode(query);
    this.setState({ query: '' });
  }

  render() {
    return(
      <form onSubmit={this.handleSubmit} className='Search'>
        <input
          value={this.state.query}
          onChange={this.handleChange}
          placeholder='search for a city'
          className='Search--input'
        />
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  forwardGeocode: (query) => dispatch(forwardGeocode(query))
});

export default connect(null, mapDispatchToProps)(Search);

Search.propTypes = {
  forwardGeocode: PropTypes.func
}