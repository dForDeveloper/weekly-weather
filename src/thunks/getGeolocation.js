import { setCoordinates, setError, toggleLoading } from '../actions';
import { reverseGeocode } from './reverseGeocode';
import { getWeather } from './getWeather';

export const getGeolocation = ({ latitude, longitude }, history) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try{
      dispatch(setCoordinates({ latitude, longitude }));
      await dispatch(reverseGeocode({ latitude, longitude }, history));
      await dispatch(getWeather({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(toggleLoading(false));
  }
}