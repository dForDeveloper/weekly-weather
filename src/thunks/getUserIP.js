import { fetchData } from '../utils/api';
import { setCoordinates, setError, toggleLoading } from '../actions';
import { reverseGeocode } from '../thunks/reverseGeocode';
import { getWeather } from '../thunks/getWeather';

export const getUserIP = (history) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try{
      const url = 'https://weekly-weather.herokuapp.com/api/v1/ip';
      const { latitude, longitude } = await fetchData(url);
      dispatch(setCoordinates({ latitude, longitude }));
      await dispatch(reverseGeocode({ latitude, longitude }, history));
      await dispatch(getWeather({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(toggleLoading(false));
  }
}