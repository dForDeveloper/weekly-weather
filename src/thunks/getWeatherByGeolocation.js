import { setCoordinates, setError, toggleLoading } from '../actions';
import { reverseGeocode } from '../thunks/reverseGeocode';
import { getWeather } from '../thunks/getWeather';

export const getWeatherByGeolocation = ({ latitude, longitude }) => {
  return (dispatch) => {
    dispatch(toggleLoading(true));
    try{
      dispatch(setCoordinates({ latitude, longitude }));
      dispatch(reverseGeocode({ latitude, longitude }));
      dispatch(getWeather({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(toggleLoading(false));
  }
}