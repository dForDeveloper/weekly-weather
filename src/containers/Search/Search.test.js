import React from 'react';
import { shallow } from 'enzyme';
import { Search, mapDispatchToProps } from './Search';
import { forwardGeocode } from '../../thunks/forwardGeocode';

jest.mock('../../thunks/forwardGeocode.js');

describe('Search', () => {
  let wrapper;
  const mockForwardGeocode = jest.fn();
  beforeEach(() => {
    wrapper = shallow(<Search forwardGeocode={mockForwardGeocode} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state('query')).toEqual('');
  });
  
  it('should set state when a change event occurs', () => {
    const mockEvent = { target: { value: 'las vegas' } };
    wrapper.find('.Search--input').simulate('change', mockEvent);
    expect(wrapper.state('query')).toEqual('las vegas');
  });
  
  it('should call forwardGeocode with the correct params on submit', () => {
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.setState({ query: 'las vegas' });
    wrapper.find('.Search').simulate('submit', mockEvent);
    expect(mockForwardGeocode).toHaveBeenCalledWith('las vegas');
    expect(wrapper.state('query')).toEqual('');
  });
});

describe('mapDispatchToProps', () => {
  it('should call dispatch with the forwardGeocode thunk', () => {
    const mockDispatch = jest.fn();
    const props = mapDispatchToProps(mockDispatch);
    const expected = forwardGeocode('las vegas');
    props.forwardGeocode('las vegas');
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});
