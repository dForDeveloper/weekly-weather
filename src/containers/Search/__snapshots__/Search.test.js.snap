// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match the snapshot 1`] = `
<form
  className="Search"
  onSubmit={[Function]}
>
  <input
    className="Search--input"
    onChange={[Function]}
    placeholder="search for a city"
    value=""
  />
</form>
`;
